### PR TITLE
gpMgmt installation (and random bits) cleanup

### PR DESCRIPTION
**Pull request by danielgustafsson**
As I mentioned in PR #8154, the installation process of `gpMgmt/` is pretty ugly as it blindly copies everything and then tries to clean it up as best as it can.

This PR is a collection of fixes, best reviewed by looking at each commit in turn. There are a few cleanup commits alongside the gist of the PR: turning `gpMgmt` into proper recursive Makefiles. Most of the code is pretty much just boilerplate as it use the infrastructure already set up for us by the upstream make system.

There are a few remaining bits left, such as ~~`gppylib`~~ and tidying up the compilation of pygresql etc. These can all be fixed easy enough, but I opted for showing the PR in this state first (partly because it's 1AM and I'm getting tired).

Updated version: gppylib is also handled now.

Ping @jmcatamney @amilkh @bradfordboyle

### CODE REVIEW

**Comment by danielgustafsson** [comment]
Disclaimer: this has only had rudimentary testing so there may be blatant bug, but it's a start..

File: gpMgmt/bin/Makefile [hunk 1]
```diff
@@ -222,8 +243,11 @@ clean distclean:
...
 	rm -rf $(PYLIB_SRC)/$(PYGRESQL_DIR)/build
 	rm -rf $(PYLIB_SRC)/subprocess32/build
-	rm -rf $(STREAM_DIR)/stream lib/stream
 	rm -rf *.pyc lib/*.pyc
-	rm -f gpcheckcatc gpexpandc
+	rm -f analyzedbc gpactivatestandbyc gpaddmirrorsc gpcheckcatc \
+		  gpcheckperfc gpcheckresgroupimplc gpchecksubnetcfgc gpconfigc \
+		  gpdeletesystemc gpexpandc gpinitstandbyc gplogfilterc gpmovemirrorsc \
+		  gppkgc gprecoversegc gpreloadc gpscpc gpsdc gpssh-exkeysc gpsshc \
+		  gpstartc gpstatec gpstopc gpsys1c minireproc
```
Comment (BaiShaoqi) [inline gpMgmt/bin/Makefile]: Following line in `gpMgmt/bin/Makefile` can be removed, they are never used and outdated.
```
#
# PSI
PSI_DIR=PSI-$(PSI_VERSION)
 psi:
	@echo "--- psi"
ifeq "$(findstring $(BLD_ARCH),aix7_ppc_64 )" ""
	cd $(PYLIB_SRC)/ && $(TAR) xzf $(PSI_DIR).tar.gz
	cd $(PYLIB_SRC)/$(PSI_DIR)/ && CC="$(CC)" python setup.py build
	cp -r $(PYLIB_SRC)/$(PSI_DIR)/build/lib.*/psi $(PYLIB_DIR)
endif
```

**Comment by BaiShaoqi** [comment]
give a salute !

File: gpMgmt/bin/Makefile (hunk 1 above)
Comment (danielgustafsson) [reply gpMgmt/bin/Makefile]: Thanks, fixed in a rebased version.

**Comment by danielgustafsson** [comment]
One important thing to note here is that this PR specifically avoids installing any .pyc files, for two reasons: a) some files are useless from the start as they come from `gpMgmt/` but we edit the installed files with the version info making the .pyc outdate; b) the files are compiled in the packaging step which overwrites any existing .pyc files anyways.

**Comment by amilkh** [comment]
Thanks for tagging me, but let me defer to the experts on this part of the code, e.g. @d @jimmyyih

**Comment by amilkh** [comment]
I will point out though that the CI is failing and it would be good to have some TDD for these changes, although parts of that may end up falling into [greenplum-database-release](https://github.com/greenplum-db/greenplum-database-release), e.g. that no .pyc files or hidden .* files are present in bin_gpdb.

**Comment by danielgustafsson** [comment]
> I will point out though that the CI is failing

It's failing on an isolation test which I can't see to be related.

> and it would be good to have some TDD for these changes, 

Testing this will be a manual process like this:

1. Wipe out the installation in `${prefix}`
1. `make install` on master, then save the resulting tree `cd ${prefix} && find . > /tmp/master.txt`
2. `make install` on this branch, then save the resulting tree `cd ${prefix} && find . > /tmp/patched.txt`
3. Compare the files, all actual production files should be in both, patched.txt should have nothing extra

**Comment by bradfordboyle** [comment]
I wanted to see what effect this would have on the result bin_gpdb tar archive:

```diff
--- server-rc-7.0.0-alpha.0+dev.543.gdd053277d4-ubuntu18.04_x86_64.tar.gz	2019-07-19 19:37:05.192873771 +0000
+++ server-rc-7.0.0-alpha.0+dev.554.g6129ba148c-ubuntu18.04_x86_64.tar.gz	2019-07-19 19:37:05.192873771 +0000
@@ -1,6 +1,4 @@
 ./bin/
-./bin/.gitignore
-./bin/.rcfile
 ./bin/GPTest.pm
 ./bin/analyzedb
 ./bin/atmsort.pl
@@ -16,7 +14,6 @@
 ./bin/ecpg
 ./bin/explain.pl
 ./bin/explain.pm
-./bin/generate-greenplum-path.sh
 ./bin/get_ereport.pl
 ./bin/gpactivatestandby
 ./bin/gpaddmirrors
@@ -83,7 +80,6 @@
 ./bin/gpssh
 ./bin/gpssh-exkeys
 ./bin/gpssh_modules/
-./bin/gpssh_modules/README.md
 ./bin/gpssh_modules/__init__.py
 ./bin/gpssh_modules/__init__.pyc
 ./bin/gpssh_modules/gppxssh_wrapper.py
@@ -96,14 +92,11 @@
 ./bin/gptorment.pl
 ./bin/initdb
 ./bin/lib/
-./bin/lib/.gitignore
-./bin/lib/.gitkeep
 ./bin/lib/__init__.py
 ./bin/lib/__init__.pyc
 ./bin/lib/crashreport.gdb
 ./bin/lib/gp_bash_functions.sh
 ./bin/lib/gp_bash_version.sh
-./bin/lib/gp_bash_version.sh.in
 ./bin/lib/gpcheckcat
 ./bin/lib/gpconfigurenewsegment
 ./bin/lib/gpcreateseg.sh
@@ -121,6 +114,7 @@
 ./bin/lib/pexpect/__init__.pyc
 ./bin/lib/pexpect/pxssh.py
 ./bin/lib/pexpect/pxssh.pyc
+./bin/lib/stream
 ./bin/minirepro
 ./bin/pg_basebackup
 ./bin/pg_config
@@ -143,9 +137,6 @@
 ./bin/postmaster
 ./bin/psql
 ./bin/reindexdb
-./bin/stream/
-./bin/stream/stream
-./bin/stream/stream.o
 ./bin/vacuumdb
 ./docs/
 ./docs/cli_help/
@@ -5795,7 +5786,6 @@
 ./lib/python/__init__.py
 ./lib/python/__init__.pyc
 ./lib/python/gppylib/
-./lib/python/gppylib/README
 ./lib/python/gppylib/__init__.py
 ./lib/python/gppylib/__init__.pyc
 ./lib/python/gppylib/commands/
@@ -5812,7 +5802,6 @@
 ./lib/python/gppylib/commands/unix.py
 ./lib/python/gppylib/commands/unix.pyc
 ./lib/python/gppylib/data/
-./lib/python/gppylib/data/.gitignore
 ./lib/python/gppylib/data/7.json
 ./lib/python/gppylib/datetimeutils.py
 ./lib/python/gppylib/datetimeutils.pyc
```

Some of this differences correct (e.g., no more .gitignore, or .gitkeep) but some look like they need further investigation. In particular, "bin/stream" looks like it's been moved into `bin/lib/stream` but is missing "stream.o".

**Comment by danielgustafsson** [comment]
> I wanted to see what effect this would have on the result bin_gpdb tar archive:

Awesome, thanks for testing!
 
> Some of this differences correct (e.g., no more .gitignore, or .gitkeep) but some look like they need further investigation. In particular, "bin/stream" looks like it's been moved into `bin/lib/stream` but is missing "stream.o".

The stream diff is a tad odd, in `gpMgmt/bin/Makefile` it is copied to `gpMgmt/bin/lib/stream`, which is also where it is in the `.gitignore`. The installed files are however a result of `cp -rp`, and make little sense (the `.o` really shouldn't be installed at all and `bin/stream/stream` is just convoluted). I'm inclined to keep this as it mostly seems like a bugfix.

**Comment by bradfordboyle** [comment]
Ok, I didn't realize that `bin/stream/stream` is itself a file and not a directory. Summarizing, here are the _major_ differences that I see that _may_ be an issue:

1. `bin/stream/stream` is moved to `bin/lib/stream`
   Will the things that depend on this still work with it in its new location?
2. `bin/generate-greenplum-path.sh` has been removed
   I've used this script recently after moving a greenplum installation to another directory. Do users have an expectation of this script being available to them (esp. when installing from an RPM/DEB).
3. `bin/lib/gp_bash_version.sh.in` has been removed
   I don't know how this file is used, but it looks like it's only used by `configure` (i.e., at compile-time). It's probably safe to not include but I wanted to high-light it.

I think these are questions best answered by @greenplum-db/cluster-management.

**Comment by jmcatamney** [comment]
>     2. `bin/generate-greenplum-path.sh` has been removed
>        I've used this script recently after moving a greenplum installation to another directory. Do users have an expectation of this script being available to them (esp. when installing from an RPM/DEB).
> 
>     3. `bin/lib/gp_bash_version.sh.in` has been removed
>        I don't know how this file is used, but it looks like it's only used by `configure` (i.e., at compile-time). It's probably safe to not include but I wanted to high-light it.

Neither of these are things customers expect to use (customers shouldn't be moving files after install, and `.in` files are indeed only used by `configure`), so it's safe to leave them out.

**Comment by danielgustafsson** [comment]
> 1. `bin/stream/stream` is moved to `bin/lib/stream`
>    Will the things that depend on this still work with it in its new location?

I suggest we install the binary in a sensible location and create symlink to the old one, in case there are any dependencies anywhere.

**Comment by danielgustafsson** [comment]
The rebase moves stream back to the original location, I realized after I had said the above that we cannot symlink it to an already existing name .. so let's just stick with the legacy installpath for now

**Comment by kalensk** [comment]
> The rebase moves stream back to the original location, I realized after I had said the above that we cannot symlink it to an already existing name .. so let's just stick with the legacy installpath for now

Since 6 is a major release which is probably the best time to make breaking changes, wouldn't it make more sense to put stream in a place where we would like it now? And if we are concerned about breaking changes, can we symlink it to the old path and have a depreciation warning, and remove it in subsequent release?

**Comment by danielgustafsson** [comment]
> Since 6 is a major release which is probably the best time to make breaking changes, wouldn't it make more sense to put stream in a place where we would like it now?

If we backport this, it wont be until after we GA (keeping the changes going into 6 to a minimum), so we'll have to live with the current location for the duration of `6X_STABLE`.

> And if we are concerned about breaking changes, can we symlink it to the old path and have a depreciation warning, and remove it in subsequent release?

That's where it becomes tricky, we can't symlink it to where we want it since it results in a filename collision. It's currently located in `bin/stream/stream`, so we cannot symlink it `bin/stream` as that would collide. What we can do is put it in `bin/lib/stream` which is where I believe the original intent was to have it.

**Comment by amilkh** [comment]
> > Since 6 is a major release which is probably the best time to make breaking changes, wouldn't it make more sense to put stream in a place where we would like it now?
> 
> If we backport this, it wont be until after we GA (keeping the changes going into 6 to a minimum), so we'll have to live with the current location for the duration of `6X_STABLE`.

We wanted to get the changes which stop bundling hidden files into 6X_STABLE, as in earlier PR: https://github.com/greenplum-db/gpdb/pull/8154
Is that possible at least?

**Comment by danielgustafsson** [comment]
> What we can do is put it in `bin/lib/stream` which is where I believe the original intent was to have it.

I ended up implementing this with a symlink to the old location as that seemed like the most sensible location for this to be installed in.

Pushed a rebase for another CI run, but there is a repeatable error on fault injection that I haven't looked into yet.

**Comment by danielgustafsson** [comment]
> We wanted to get the changes which stop bundling hidden files into 6X_STABLE, as in earlier PR: #8154
> Is that possible at least?

I have limited time for figuring out the final issue with the fault injection, do you guys have any spare moments to help with that? If we can find that I guess it's doable as there should be no functional changes for the resulting package from this.

**Comment by amilkh** [comment]
> I have limited time for figuring out the final issue with the fault injection, do you guys have any spare moments to help with that? If we can find that I guess it's doable as there should be no functional changes for the resulting package from this.

Ah, I would help if I could, but maybe CM team or server team people would be better at debugging those kinds of issues...

**Comment by danielgustafsson** [comment]
Pushed a rebase on top of the Perl cleanup which really makes the gpMgmt Makefile look (mostly) pretty now.

**Comment by danielgustafsson** [comment]
Pushed to master after another re-read with tiny bits of polish. Thanks a lot to everyone for reviewing!

**Comment by bradfordboyle** [comment]
Was this every cherry-picked to 6X_STABLE?

**Comment by danielgustafsson** [comment]
> On 7 Aug 2019, at 20:20, Bradford Boyle <notifications@github.com> wrote:
> 
> Was this every cherry-picked to 6X_STABLE?

Not yet, I wanted to wait and catch the fallout from this that I expected would come (and sure enough did, for example 9f707e1e7a2caac426e3f34db00fe1448bd28d6d, 806cc27c68c3decb21a4ce64b4a4a84e7c435574 and 20b521cab102b5769e616d3b156471f7a9da539c). At this point, I don’t think we’ll see too much more though. If you want to, you are very welcome to cherry-pick this down into 6X_STABLE.

**Comment by bradfordboyle** [comment]
Just to confirm, the following commits need to be cherry-picked (newest to oldest):

9f707e1e7a2caac426e3f34db00fe1448bd28d6d
806cc27c68c3decb21a4ce64b4a4a84e7c435574
20b521cab102b5769e616d3b156471f7a9da539c
aa1640a651c7260a67715f4e35658f28c95829fa
832896ff851836b61bd6728a0f41008109d059f3
22e8d1dfe560722038ef4ad09cc344b3b6fa874d
6de21f62a2dedbabcc7d08d6187cf625d8f68d99
b5aba18b7d02d69d623c3d46b693651e2df71494
0cad6637313ab2e28c8ec2951c4ae12f68f989fb
e12a1129f1b25032ab04cef73e3a5d41caa2db07
4bfd3875649e90ec610c37d559a9ef5d857cdd04
45478fb2d9d954f3feebf36ecef9dd1bf542317c
14e2f8bc96b2227f127978ebeb4f6486d879bd9a
c41b48ab218e361ad9ab28ca9e7213f55fc8d2ed

**Comment by danielgustafsson** [comment]
> On 8 Aug 2019, at 20:55, Bradford Boyle <notifications@github.com> wrote:
> 
> Just to confirm, the following commits need to be cherry-picked (newest to oldest):

That does look about right yes. Thanks!

**Pull request closed without merging**